File: algorithm/c_algorithm/src/a.1-2_doubly_lined_list/DoublyLinkedList.c
#include "DoublyLinkedList.h"

Node* DLL_CreateNode(ElementType NewData) {
	Node* NewNode = (Node*) malloc(sizeof(Node));

	NewNode->Data = NewData;
	NewNode->PrevNode = NULL;
	NewNode->NextNode = NULL;

	return NewNode;
}

void DLL_DestroyNode(Node* Node) {
	free(Node);
}

void DLL_AppendNode(Node** Head, Node* NewNode) {
	if ((*Head) == NULL) {
		*Head = NewNode;
	} else {
		Node* Tail = (*Head);
		while (Tail->NextNode != NULL) {
			Tail = Tail->NextNode;
		}
		Tail->NextNode = NewNode;
		NewNode->PrevNode = Tail;
	}
}

void DLL_InsertAfter(Node* Current, Node* NewNode) {
	NewNode->NextNode = Current->NextNode;
	NewNode->PrevNode = Current;

	if (Current->NextNode != NULL) {
		Current->NextNode->PrevNode = NewNode;
	}
	Current->NextNode = NewNode;
}

void DLL_RemoveNode(Node** Head, Node* Remove) {
	if (*Head == Remove) {
		*Head = Remove->NextNode;
		if ((*Head) != NULL) {
			(*Head)->PrevNode = NULL;
		}
	} else {
		Node* Temp = Remove;

		if (Remove->PrevNode != NULL) {
			Remove->PrevNode->NextNode = Temp->NextNode;
		}
		if (Remove->NextNode != NULL) {
			Remove->NextNode->PrevNode = Temp->PrevNode;
		}
	}
	Remove->NextNode = NULL;
	Remove->PrevNode = NULL;
}

Node* DLL_GetNodeAt(Node* Head, int Location) {
	Node* Current = Head;
	while(Current != NULL && (--Location >=0)){
		Current = Current->NextNode;
	}
	return Current;
}


int DLL_GetNodeCount(Node* Head){
	unsigned int Count = 0;
	Node* Current = Head;
	while(Current != NULL){
		Count++;
		Current = Current->NextNode;
	}
	return Count;
}

File: algorithm/Java_Algorithm/src/a_linked_list/DoublyLinkedList.java
package a_linked_list;

public class DoublyLinkedList<T> {
	private int size;
	private Node head;
	
	private DoublyLinkedList() {
		this.size = 0;
		this.head = null; 
	}
	
	class Node{
		private T value;
		private Node nextNode;
		private Node prevNode;
		
		public Node(T value) {
			this.value = value;
			this.nextNode = null;
			this.prevNode = null;
		}
	}
	
	private void add(T value) {
		Node newNode = new Node(value);
		if(this.head == null) {
			this.head = newNode;
		}else {
			Node tail = head;
			while(tail.nextNode != null) {
				tail = tail.nextNode;
			}
			tail.nextNode = newNode;
			newNode.prevNode = tail;
		}
		this.size++;
	}
	
	private T get(int index) {
		Node tail = this.head;
		if(index < 0 || index >= size ) {
			throw new NullPointerException("데이터 없음 : " + index);
		}else {
			while((tail.nextNode != null) && (--index >= 0)) {
				tail = tail.nextNode;
			}
		}
		return tail.value;
	}
	
	private void remove(int index) {
		Node tail = this.head;
		if(index < 0 || index >= size ) {
			throw new NullPointerException("데이터 없음 : " + index);
		}else {
			while(tail.nextNode != null && (--index >=0)) {
				tail = tail.nextNode;
			}
			Node current = tail;
			if(null != current.prevNode) {
				tail.prevNode.nextNode = current.nextNode;
			}
			if(null != current.nextNode) {
				tail.nextNode.prevNode = current.prevNode;
			}
			
			current.nextNode = null;
			current.prevNode = null;
			current.value = null;
			this.size--;
		}
	}
	
	private void update(int index, T value) {
		Node tail = this.head;
		if(index < 0 || index >= size ) {
			throw new NullPointerException("데이터 없음 : " + index);
		}else{
			while(tail.nextNode != null && (--index >= 0)) {
				tail = tail.nextNode;
			}
			tail.value = value;
		}
	}
	
	private void print() {
		for(int i = 0; i < this.size; i++) {
			System.out.println(this.get(i));
		}
		System.out.println("print() ---- end");
	}
	public static void main(String args[]) {
		DoublyLinkedList<String> list = new DoublyLinkedList<>();
		list.add("1");
		list.add("2");
		list.add("3");
		
		list.print();
		
		
		list.remove(2);
		list.update(0,"6");
//		list.update(1,"6");
//		list.update(2,"6");
		list.print();
	}
	
}
